extensions/wikia/AdEngine3/AdEngine3ApiController.class.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+class AdEngine3ApiController extends WikiaController
+{
+	const DEFAULT_TEMPLATE_ENGINE = WikiaResponse::TEMPLATE_ENGINE_MUSTACHE;
+
+	public function getRecCode()
+	{
+		global $wgUser;
+
+		$this->response->setContentType('text/javascript');
+		$this->response->setCachePolicy(WikiaResponse::CACHE_PUBLIC);
+		$this->response->setCacheValidity(WikiaResponse::CACHE_LONG);
+
+		if ($wgUser->isAnon()) {
+			$resourceLoader = null;
+			$type = $this->request->getVal('type', 'rec');
+
+			switch ($type) {
+				case 'bt':
+					$resourceLoader = new ResourceLoaderAdEngine3BTCode();
+					break;
+
+				case 'hmd':
+					$resourceLoader = new ResourceLoaderAdEngine3HMDCode();
+					break;
+			}
+
+			if ($resourceLoader) {
+				$resourceLoaderContext = new ResourceLoaderContext(new ResourceLoader(), new FauxRequest());
+				$this->response->setBody($resourceLoader->getScript($resourceLoaderContext));
+			} else {
+				$this->response->setBody('tut22ej');
+			}
+		} else {
+			$this->response->setBody('');
```

Comment:
What about users who are logged in AND want to receive ads? "We can ignore them" is also a good answer here.

Reply:
Yes, we can ignore them. It's a business decision to disable AdBlock recovery for logged users.